src/core/qgstiles.cpp

```diff
@@ -259,8 +260,8 @@ double QgsTileMatrixSet::scaleToZoom( double scale ) const
 
   if ( zoomUnder < 0 )
     return zoomOver;
-  if ( zoomOver < 0 )
-    return zoomUnder;
+  if ( zoomOver < 0 ) // allow overzooming, so the styling is applied correctly
+    return QgsVectorTileUtils::scaleToZoom( scale );
```

Comment:
This still won't work nicely for the esri tile matrices -- when overzooming we'lll suddenly get the tile zoom doubling which is done by QgsVectorTileUtils::scaleToZoom but isn't desirable for ESRI tiles.

Instead we need to start from the highest available zoom matrix scale, and keep doubling the scale here until we hit the desired range. That'll ensure correct results for esri/non esri tiles.